### DISCUSSION TITLE
DDR lords default to following Luxor

### DISCUSSION THREAD

**Discussion by derfsss** [ChilliHugger/The-Lords-Of-Midnight#247] [Code Algorithm Questions] [answered]
Is it correct that any lords that do not have a liege e.g. Imgorad the Barbarian, have a chance to go to Luxor's location ?

I can understand that rule for lords where their foe is Luxor, but its currently possible to have Imgorad or Carormand hunting you down straight away - possibly with several lords following them.

**Comment by IcemarkUK** [comment]
CH_IMGORAD starts with the orders to head home. He could choose to follow his liege's orders CH_CARANGRANE

CH_CARORMAND also has orders to head home.  He could choose to follow his liege's orders CH_LORANGRARG

Neither liege starts with any interest in Luxor.

The only real reason for either to attack Luxor is for them to be in proximity.

Do you have the new AI rules enabled? `Upgrade to Shareth Army AI` as this would make it more likely that a lone lord would attack.

**Reply by derfsss** [reply]
https://github.com/ChilliHugger/The-Lords-Of-Midnight/blob/4c61a33a13d0ae1c30b2183461b448bd1de14323/main/midnight/src/tme/scenarios/ddr/ddr_character.cpp#L1129

CH_IMGORAD has no liege (CH_CARANGRANE is the foe) but it still stands.

CH_IMGORAD starts at his home city, and say that based on the first random roll he gets a new order to follow liege. He has no liege so defaults to Luxor as below.

https://github.com/ChilliHugger/The-Lords-Of-Midnight/blob/4c61a33a13d0ae1c30b2183461b448bd1de14323/main/midnight/src/tme/scenarios/ddr/ddr_character.cpp#L1132

If the random roll each night is always 5+ he will go south to Luxor and end up fighting - possibly bringing other lords whose liege is CH_IMGORAD.

Could this not be true for all characters without a liege, or even characters without a foe and the new order is to follow their foe?

Am I reading this correctly or is this part of the code something else?

Thanks for looking at this.

**Reply by FrozenEmpire** [reply]
Seems an entirely plausible sequence of events based on my understanding. 

whether thats original DDR behaviour only Chris can confirm :)

**Reply by IcemarkUK** [reply, marked as answer]
> https://github.com/ChilliHugger/The-Lords-Of-Midnight/blob/4c61a33a13d0ae1c30b2183461b448bd1de14323/main/midnight/src/tme/scenarios/ddr/ddr_character.cpp#L1129
> 
> CH_IMGORAD has no liege (CH_CARANGRANE is the foe) but it still stands.
> 
> CH_IMGORAD starts at his home city, and say that based on the first random roll he gets a new order to follow liege. He has no liege so defaults to Luxor as below.

Sorry, you are entirely right, [I was misreading the data on my device...](https://github.com/ChilliHugger/The-Lords-Of-Midnight/blob/4c61a33a13d0ae1c30b2183461b448bd1de14323/ddr/database/characters.tsv#L103)

You are reading it entirely right. What you describe could essentially happen. And to be fair, on the original, if you leave Luxor at the Gate of Varenorn - stick him in the tunnel - and see how many barbarians turn up!

This part of the problem of the "AI" in DDR, it's really not that good. Too many random numbers. Now, random numbers as we know are always problematic because what the game actually wants is probability. Also the random no generator in my version is not the RNG on the original (which was based on reading current memory address pointed to by the hardware refresh register).

I will however take a quick look at the original ASM, just to be sure.

**Reply by IcemarkUK** [reply]
So I took a look at the ASM and I am happy with my interpretation. [Here is the original for you to look over.](https://gist.github.com/IcemarkUK/8216a749fb7373fb8983a3793659747c#file-whatischaracterdoing-z80)

The only thing I am a concerned about would be the use of GetCharsNextLiege. Both selecting the foe/liege is a loop. I originally wrote it like that and you can see where I commented out - due to an endless loop. I think I might need to revisit that code.

**Reply by derfsss** [reply]
thats fair - I expect i'm just experiencing the different RNG in action as it may just be more probable now than in the past!
